org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/id/IdInventory.java

```diff
@@ -0,0 +1,86 @@
+import org.eclipse.scout.rt.platform.CreateImmediately;
+import org.eclipse.scout.rt.platform.inventory.ClassInventory;
+import org.eclipse.scout.rt.platform.inventory.IClassInfo;
+import org.eclipse.scout.rt.platform.util.Assertions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+@CreateImmediately // validate id types on startup
+public class IdInventory {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IdInventory.class);
+
+  protected final Map<String, Class<? extends IId>> m_nameToClassMap = new HashMap<>();
+  protected final Map<Class<? extends IId>, String> m_classToNameMap = new HashMap<>();
+
+  @PostConstruct
+  protected void createClassCache() {
+    for (IClassInfo classInfo : ClassInventory.get().getKnownAnnotatedTypes(IdTypeName.class)) {
+      String typeName = (String) classInfo.getAnnotationValue(IdTypeName.class, "value");
+      Assertions.assertNotNullOrEmpty(typeName, "Invalid value for @{} on {} (must not be null or empty)", IdTypeName.class.getSimpleName(), classInfo.resolveClass().getName());
+      try {
+        Class<? extends IId> idClass = classInfo.resolveClass().asSubclass(IId.class);
+        Class<? extends IId> registeredIdClass = m_nameToClassMap.put(typeName, idClass);
+        String registeredTypeName = m_classToNameMap.put(idClass, typeName);
+        checkDuplicateIdTypeNames(idClass, typeName, registeredIdClass, registeredTypeName);
+      }
+      catch (@SuppressWarnings("squid:S1166") ClassCastException e) {
+        LOG.warn("Class {} is annotated with @{} but does not implement {}. Skipping class.", classInfo.resolveClass().getName(), IdTypeName.class.getSimpleName(), IId.class.getName());
+      }
+    }
+    LOG.debug("Registered {} id types", m_nameToClassMap.size());
+  }
+
+  /**
+   * Checks for classes with the same {@link IdTypeName} annotation values.
+   */
+  protected void checkDuplicateIdTypeNames(Class<?> clazz, String typeName, Class<?> existingClass, String existingName) {
+    Assertions.assertNull(existingClass, "{} and {} have the same type name '{}'. Use an unique @{} annotation value.", clazz, existingClass, typeName, IdTypeName.class.getSimpleName());
+    Assertions.assertNull(existingName, "{} is annotated with @{} value '{}', but was already registered with type name '{}'. Register each class only once.", clazz, IdTypeName.class.getSimpleName(), typeName, existingName);
+  }
+
+  /**
+   * @return the type name of the id class as defined by the {@link IdTypeName} annotation or <code>null</code> if the
+   *         annotation is not present.
+   */
+  public String getTypeName(Class<? extends IId> idClass) {
+    return m_classToNameMap.get(idClass);
+  }
+
+  /**
+   * @return the type name of the {@link IId} as defined by the {@link IdTypeName} annotation or <code>null</code> if
+   *         the annotation is not present.
+   */
+  public String getTypeName(IId id) {
```

Comment:
Define protected and remove after IdExternalFormatter was removed. For an inventory, providing the access via Class is sufficient, no need to access by instance.

Reply:
As discussed, method is used variously through code. Added javadoc.